semgrep-core/src/core/ast/AST_generic.ml

```diff
@@ -476,6 +476,13 @@ and expr_kind =
   (* less: could be in Special, but pretty important so I've lifted them here*)
   | Ref of tok (* &, address of *) * expr
   | DeRef of tok (* '*' in C, '!' or '<-' in OCaml, ^ in Reason *) * expr
+  (* In some rare cases, we need to keep the parenthesis around an expression
+   * otherwise in autofix semgrep could produce incorrect code. For example,
+   * in Go a cast int(3.0) requires the parenthesis.
+   * alt: change cast to take a expr bracket, but this is used only for Go.
+   * Note that this data structure is really becoming more a CST than an AST.
+   *)
+  | ParenExpr of expr bracket
```

Comment:
I don't know how I feel about adding this. Right now it seems like we aren't going into the direction of saving useless parentheses? (e.g. `((x + 1)) + 2`) But adding `ParenExpr` instead of adding tokens to `Cast` seems to imply we want to go that direction. 

The generic AST is already kind of a CST, so maybe that is the direction we want to go for autofix, but I'm not convinced yet. It seems to me like our problem with the "string replace" method of autofix is that we rely on storing tokens for expression ranges (instead of saving a range with each AST element). Not that I think that's necessarily the correct approach, either. But I don't think we've made a decision

For now, it looks like adding tokens to `Cast` would also solve this problem and adhere closer to our current philosophy? Though I guess it's annoying because languages like C put the parentheses around the type.

If we do go on with adding a `ParenExpr`, this issue also needs it: https://github.com/returntocorp/semgrep/issues/2051

Reply:
I totally agree with you. I am not super convinced either. I think of it as an experiment ...